src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/BindToStreamBindingProvider.cs

```diff
@@ -0,0 +1,518 @@
+        private readonly FileAccess _access; // Which direction this rule applies to. Can be R, W, or  RW
+        private readonly INameResolver _nameResolver;
+        private readonly IConverterManager _converterManager;
+
+        private readonly FuncAsyncConverter<TAttribute, Stream> _builder;
+
+        public BindToStreamBindingProvider(FuncAsyncConverter<TAttribute, Stream> builder, FileAccess access, INameResolver nameResolver, IConverterManager converterManager)
+        {
+            _nameResolver = nameResolver;
+            _converterManager = converterManager;
+            _access = access;
+
+            _builder = builder;
+        }
+
+        public Type GetDefaultType(Attribute attribute, FileAccess access, Type requestedType)
+        {            
+            if (attribute is TAttribute)
+            {
+                if (access == FileAccess.Read)
+                {
+                    if (CanRead(this._access))
+                    {
+                        return typeof(Stream);
+                    }
+                }
+                if (access == FileAccess.Write)
+                {
+                    if (CanWrite(this._access))
+                    {
+                        return typeof(Stream);
+                    }    
+                }
+                // Read write stream area not supported. 
+            }
+
+            return null;
+        }
+
+        public IEnumerable<BindingRule> GetRules()
+        {
+            foreach (var type in new Type[]
+            {
+                typeof(Stream),
+                typeof(TextReader),
+                typeof(TextWriter),
+                typeof(string),
+                typeof(byte[]),
+                typeof(string).MakeByRefType(),
+                typeof(byte[]).MakeByRefType()
+            })
+            {
+                yield return new BindingRule
+                {
+                    SourceAttribute = typeof(TAttribute),
+                    UserType = OpenType.FromType(type)
+                };
+            }
+        }
+
+        static private void VerifyAccessOrThrow(FileAccess? declaredAccess, bool isRead)
+        {
+            // Verify direction is compatible with the attribute's direction flag. 
+            if (declaredAccess.HasValue)
+            {
+                string errorMsg = null;
+                if (isRead)
+                {
+                    if (!CanRead(declaredAccess.Value))
+                    {
+                        errorMsg = "Read";
+                    }
+                }
+                else
+                {
+                    if (!CanWrite(declaredAccess.Value))
+                    {
+                        errorMsg = "Write";
+                    }
+                }
+                if (errorMsg != null)
+                {
+                    throw new InvalidOperationException($"The parameter type is a '{errorMsg}' binding, but the Attribute's access type is '{declaredAccess}'");
```

Comment:
I wonder if in error messages, rather than say things like "the Attribute's access" we shouldn't avoid mentioning attributes. This will yield better user messages for AF (often they're not dealing with attributes, e.g. Node).

Reply:
Can you suggest a wording?  "attribute --> "declared access". 
"...  but the binding's declared access is ..."

---
In reply to: [150924524](https://github.com/Azure/azure-webjobs-sdk/pull/1422#discussion_r150924524) [](ancestors = 150924524)